python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py

```diff
@@ -379,3 +386,214 @@ def get_logs_for_run(
+    timestamp is the timestamp of the oldest entry (run or backfill).
+
+    If the run/backfill cursor is None, that means that no runs/backfills have been returned yet and querying
+    should begin at the start of the table. Once all runs/backfills in the table have been returned, the
+    corresponding cursor should still be set to the id of the last run/backfill returned.
+
+    The timestamp is used for the following case. If a deployment has 20 runs and 0 backfills, and a query is
+    made for 10 Runs Feed entries, the first 10 runs will be returned. At this time, the run_cursor will be an id,
+    and the backfill_cursor will be None. Then a backfill is created. If a second query is made for 10 Runs Feed entries
+    the newly created backfill will get included in the list, even though it should be included on the first page by time
+    order. To prevent this, the timestamp is used to ensure that all returned entires are older than the entries on the
+    previous page.
+    """
+
+    run_cursor: Optional[str]
+    backfill_cursor: Optional[str]
+    timestamp: Optional[float]
+
+    def to_string(self) -> str:
+        return f"{self.run_cursor if self.run_cursor else ''}{_DELIMITER}{self.backfill_cursor if self.backfill_cursor else ''}{_DELIMITER}{self.timestamp if self.timestamp else ''}"
+
+    @staticmethod
+    def from_string(serialized: Optional[str]):
+        if serialized is None:
+            return RunsFeedCursor(
+                run_cursor=None,
+                backfill_cursor=None,
+                timestamp=None,
+            )
+        parts = serialized.split(_DELIMITER)
+        if len(parts) != 3:
+            raise DagsterInvariantViolationError(f"Invalid cursor for querying runs: {serialized}")
+
+        return RunsFeedCursor(
+            run_cursor=parts[0] if parts[0] else None,
+            backfill_cursor=parts[1] if parts[1] else None,
+            timestamp=float(parts[2]) if parts[2] else None,
+        )
+
+
+def _fetch_runs_not_in_backfill(
+    instance: DagsterInstance,
+    cursor: Optional[str],
+    limit: Optional[int],
+    created_before: Optional[float] = None,
+) -> Sequence[RunRecord]:
+    """Fetches limit RunRecords that are not part of a backfill and were created before a given timestamp."""
+    runs_filter = (
+        RunsFilter(created_before=datetime_from_timestamp(created_before))
+        if created_before
+        else None
+    )
+    if limit is None:
+        new_runs = instance.get_run_records(cursor=cursor, filters=runs_filter)
+        return [run for run in new_runs if run.dagster_run.tags.get(BACKFILL_ID_TAG) is None]
+
+    runs = []
+    while len(runs) < limit:
+        new_runs = instance.get_run_records(limit=limit, cursor=cursor, filters=runs_filter)
+        if len(new_runs) == 0:
+            return runs
+        cursor = new_runs[-1].dagster_run.run_id
+        runs.extend([run for run in new_runs if run.dagster_run.tags.get(BACKFILL_ID_TAG) is None])
```

Comment:
nit: maybe a summary comment about having to loop to fill discarded backfill runs